File: eclipse_plugin/src/com/sina/weibo/agent/eclipse/core/ExtensionSocketServer.java
/*
 * SPDX-FileCopyrightText: 2025 Weibo, Inc.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.sina.weibo.agent.eclipse.core;

import java.io.IOException;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;

import com.sina.weibo.agent.eclipse.Activator;
import com.sina.weibo.agent.eclipse.ipc.NodeSocket;

/**
 * Socket server for extension host communication.
 * Listens for connections from the Node.js extension host process.
 */
public class ExtensionSocketServer {

    /** Server socket */
    private ServerSocket serverSocket;

    /** Server port */
    private int port;

    /** Running flag */
    private final AtomicBoolean running = new AtomicBoolean(false);

    /** Disposed flag */
    private final AtomicBoolean disposed = new AtomicBoolean(false);

    /** Executor for accept loop */
    private ExecutorService acceptExecutor;

    /** Client connection handler */
    private Consumer<NodeSocket> connectionHandler;

    /** Current project path */
    private String projectPath;

    /** Current connected client */
    private NodeSocket currentClient;

    /**
     * Create a new ExtensionSocketServer.
     */
    public ExtensionSocketServer() {
        Activator.logInfo("ExtensionSocketServer created");
    }

    /**
     * Start the server.
     *
     * @param projectPath the project path
     * @return the port number, or -1 if failed
     */
    public int start(String projectPath) {
        if (running.get()) {
            Activator.logWarning("Server already running on port " + port);
            return port;
        }

        this.projectPath = projectPath;

        try {
            // Create server socket on a random available port
            serverSocket = new ServerSocket(0, 50, InetAddress.getLoopbackAddress());
            port = serverSocket.getLocalPort();

            Activator.logInfo("ExtensionSocketServer started on port " + port);

            // Start accept loop
            acceptExecutor = Executors.newSingleThreadExecutor(r -> {
                Thread t = new Thread(r, "ExtensionSocketServer-Accept");
                t.setDaemon(true);
                return t;
            });

            running.set(true);
            acceptExecutor.submit(this::acceptConnections);

            return port;

        } catch (IOException e) {
            Activator.logError("Failed to start ExtensionSocketServer", e);
            return -1;
        }
    }

    /**
     * Accept loop for incoming connections.
     */
    private void acceptConnections() {
        while (running.get() && !disposed.get()) {
            try {
                Socket clientSocket = serverSocket.accept();
                Activator.logInfo("Client connected from " + clientSocket.getRemoteSocketAddress());

                // Wrap in NodeSocket
                NodeSocket nodeSocket = new NodeSocket(clientSocket, "ExtHostClient");
                currentClient = nodeSocket;

                // Notify connection handler
                if (connectionHandler != null) {
                    connectionHandler.accept(nodeSocket);
                }

                // Start receiving data
                nodeSocket.startReceiving();

            } catch (IOException e) {
                if (running.get() && !disposed.get()) {
                    Activator.logError("Error accepting connection", e);
                }
            }
        }
    }

    /**
     * Stop the server.
     */
    public void stop() {
        if (!running.getAndSet(false)) {
            return;
        }

        Activator.logInfo("Stopping ExtensionSocketServer...");

        // Close current client
        if (currentClient != null) {
            currentClient.dispose();
            currentClient = null;
        }

        // Close server socket
        if (serverSocket != null) {
            try {
                serverSocket.close();
            } catch (IOException e) {
                Activator.logError("Error closing server socket", e);
            }
            serverSocket = null;
        }

        // Shutdown executor
        if (acceptExecutor != null) {
            acceptExecutor.shutdownNow();
            acceptExecutor = null;
        }

        Activator.logInfo("ExtensionSocketServer stopped");
    }

    /**
     * Connect to a debug host (for development).
     *
     * @param host the host address
     * @param port the port number
     * @return true if connected
     */
    public boolean connectToDebugHost(String host, int port) {
        try {
            Activator.logInfo("Connecting to debug host at " + host + ":" + port);
            
            Socket socket = new Socket(host, port);
            NodeSocket nodeSocket = new NodeSocket(socket, "DebugHost");
            currentClient = nodeSocket;

            // Notify connection handler
            if (connectionHandler != null) {
                connectionHandler.accept(nodeSocket);
            }

            // Start receiving data
            nodeSocket.startReceiving();

            Activator.logInfo("Connected to debug host");
            return true;

        } catch (IOException e) {
            Activator.logError("Failed to connect to debug host", e);
            return false;
        }
    }

    /**
     * Set the connection handler.
     *
     * @param handler the handler to call when a client connects
     */
    public void setConnectionHandler(Consumer<NodeSocket> handler) {
        this.connectionHandler = handler;
    }

    /**
     * Get the server port.
     *
     * @return the port or -1 if not started
     */
    public int getPort() {
        return serverSocket != null ? port : -1;
    }

    /**
     * Check if the server is running.
     *
     * @return true if running
     */
    public boolean isRunning() {
        return running.get();
    }

    /**
     * Get the current client socket.
     *
     * @return the current client or null
     */
    public NodeSocket getCurrentClient() {
        return currentClient;
    }

    /**
     * Get the project path.
     *
     * @return the project path
     */
    public String getProjectPath() {
        return projectPath;
    }

    /**
     * Dispose the server and release resources.
     */
    public void dispose() {
        if (disposed.getAndSet(true)) {
            return;
        }

        stop();
        Activator.logInfo("ExtensionSocketServer disposed");
    }
}

File: eclipse_plugin_neo/src/main/kotlin/com/sina/weibo/agent/eclipse/neo/core/ExtensionSocketServer.kt
// SPDX-FileCopyrightText: 2025 Weibo, Inc.
//
// SPDX-License-Identifier: Apache-2.0

package com.sina.weibo.agent.eclipse.neo.core

// import com.intellij.openapi.Disposable
// import com.intellij.openapi.diagnostic.Logger
// import com.intellij.openapi.project.Project
import java.net.ServerSocket
import java.net.Socket
import java.io.IOException
import java.util.concurrent.ConcurrentHashMap
import kotlin.concurrent.thread

/**
 * Extension process Socket server
 * Used to establish communication with extension process
 */
interface ISocketServer {
    fun start(projectPath: String = ""): Any?
    fun stop()
    fun isRunning(): Boolean
    fun dispose()
}

class ExtensionSocketServer() : ISocketServer {
    private fun logger(level: String, message: String, e: Exception? = null) {
        println("[$level] ExtensionSocketServer: $message")
        e?.printStackTrace()
    }

    // Server socket
    private var serverSocket: ServerSocket? = null

    // Connected client managers
    private val clientManagers = ConcurrentHashMap<Socket, Any>() // TODO: Replace Any with ExtensionHostManager

    // Server thread
    private var serverThread: Thread? = null

    // Current project path
    private var projectPath: String = ""

    // Whether running
    @Volatile
    private var isRunning = false

    // lateinit var project: Project

    /**
     * Start Socket server
     * @param projectPath Current project path
     * @return Server port, -1 if failed
     */
    override fun start(projectPath: String): Int {
        if (isRunning) {
            logger("INFO", "Socket server is already running")
            return serverSocket?.localPort ?: -1
        }

        this.projectPath = projectPath

        try {
            // Use 0 to indicate random port assignment
            serverSocket = ServerSocket(0)
            val port = serverSocket?.localPort ?: -1

            if (port <= 0) {
                logger("ERROR", "Failed to get valid port for socket server")
                return -1
            }

            isRunning = true
            logger("INFO", "Starting socket server on port: $port")

            // Start the thread to accept connections
            serverThread = thread(start = true, name = "ExtensionSocketServer") {
                acceptConnections()
            }

            return port
        } catch (e: Exception) {
            logger("ERROR", "Failed to start socket server", e)
            stop()
            return -1
        }
    }

    /**
     * Stop Socket server
     */
    override fun stop() {
        if (!isRunning) {
            return
        }

        isRunning = false
        logger("INFO", "Stopping socket server")

        // Close all client managers
        clientManagers.forEach { (_, manager) ->
            try {
                // (manager as ExtensionHostManager).dispose()
            } catch (e: Exception) {
                logger("WARN", "Failed to dispose client manager", e)
            }
        }
        clientManagers.clear()

        // Close the server
        try {
            serverSocket?.close()
        } catch (e: IOException) {
            logger("WARN", "Failed to close server socket", e)
        }

        // Interrupt the server thread
        serverThread?.interrupt()
        serverThread = null
        serverSocket = null

        logger("INFO", "Socket server stopped")
    }

    /**
     * Thread function for accepting connections
     */
    private fun acceptConnections() {
        val server = serverSocket ?: return

        logger("INFO", "Socket server started, waiting for connections..., tid: ${Thread.currentThread().id}")

        while (isRunning && !Thread.currentThread().isInterrupted) {
            try {
                val clientSocket = server.accept()
                logger("INFO", "New client connected from: ${clientSocket.inetAddress.hostAddress}")

                clientSocket.tcpNoDelay = true // Set no delay

                // Create extension host manager
                // val manager = ExtensionHostManager(clientSocket, projectPath,project)
                // clientManagers[clientSocket] = manager

                // handleClient(clientSocket, manager)
            } catch (e: IOException) {
                if (isRunning) {
                    logger("ERROR", "Error accepting client connection", e)
                } else {
                    // IOException is thrown when ServerSocket is closed, this is normal
                    logger("INFO", "Socket server closed")
                    break
                }
            } catch (e: InterruptedException) {
                // Thread interrupted, this is normal
                logger("INFO", "Socket server thread interrupted")
                break
            } catch (e: Exception) {
                logger("ERROR", "Unexpected error in accept loop", e)
                if (isRunning) {
                    try {
                        // Retry after short delay
                        Thread.sleep(1000)
                    } catch (ie: InterruptedException) {
                        // Thread interrupted, server is shutting down
                        logger("INFO", "Socket server thread interrupted during sleep")
                        break
                    }
                }
            }
        }

        logger("INFO", "Socket accept loop terminated")
    }

    /**
     * Handle client connection
     */
    private fun handleClient(clientSocket: Socket, manager: Any /* ExtensionHostManager */) {
        // ... (Implementation will be added once ExtensionHostManager is ported)
    }

    /**
     * Get current port
     */
    fun getPort(): Int {
        return serverSocket?.localPort ?: -1
    }

    /**
     * Whether running
     */
    override fun isRunning(): Boolean {
        return isRunning
    }

    /**
     * Resource cleanup
     */
    override fun dispose() {
        stop()
    }

    /**
     * Connect to debug host
     * @param host Debug host address
     * @param port Debug host port
     * @return Whether connection is successful
     */
    fun connectToDebugHost(host: String, port: Int): Boolean {
        if (isRunning) {
            logger("INFO", "Socket server is already running, stopping first")
            stop()
        }

        try {
            logger("INFO", "Connecting to debug host at $host:$port")

            // Directly connect to the specified address and port
            val clientSocket = Socket(host, port)
            clientSocket.tcpNoDelay = true // Set no delay

            isRunning = true

            // Create extension host manager
            // val manager = ExtensionHostManager(clientSocket, projectPath,project)
            // clientManagers[clientSocket] = manager

            // Start connection handling in background thread
            // thread(start = true, name = "DebugHostHandler") {
            //     handleClient(clientSocket, manager)
            // }

            logger("INFO", "Successfully connected to debug host at $host:$port")
            return true
        } catch (e: Exception) {
            logger("ERROR", "Failed to connect to debug host at $host:$port", e)
            stop()
            return false
        }
    }
}
